### PR TITLE
Fixed #27709 -- Fixed get_for_models() for proxies with an empty cache.

### PR DESCRIPTION
**Pull request by charettes**
Thanks Peter Inglesby for the report and tests.

### CODE REVIEW

**Comment by timgraham** [comment]
Do you think the code is clear enough without any additional comments?

**Comment by charettes** [comment]
@timgraham I do now that I've spent some time dealing with it but given this issue was in there for so long some extra comments wouldn't hurt.

**Comment by charettes** [comment]
Adding comments made me realize the code had a bug with contenttypes creation. I added an test and fix the issue so I'll proceed with merge once the build goes green.

File: django/contrib/contenttypes/models.py [hunk 1]
```diff
@@ -65,41 +67,43 @@ def get_for_models(self, *models, **kwargs):
...
                 model = ct.model_class()
-                if model._meta in needed_opts:
+                opts_models = needed_opts.pop(ct.model_class()._meta, [])
+                for model in opts_models:
                     results[model] = ct
-                    needed_opts.remove(model._meta)
                 self._add_to_cache(self.db, ct)
-        for opts in needed_opts:
-            # These weren't in the cache, or the DB, create them.
+        # These weren't in the cache, or the DB, create them.
```
Comment (timgraham) [inline django/contrib/contenttypes/models.py]: You could rewrite this as: "Create content types that weren't in the cache or DB."

**Pull request merged**